### PR TITLE
Added partition-offset mapper to kafka

### PR DESCRIPTION
**Pull request by gurbuzali**

### CODE REVIEW

File: hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/connector/kafka/ReadKafkaP.java [hunk 1]
```diff
@@ -145,33 +155,62 @@ public void close() {
...
+                                                         Function<byte[], V> deserializeValue) {
+        return readKafka(zkAddress, groupId, topicId, brokerConnectionString, deserializeKey, deserializeValue,
+                CURRENT_POSITION_OFFSET_MAPPER);
+    }
+
+    /**
+     * Returns a meta-supplier of processors that consume a kafka topic and emit
+     * items from it as {@code Map.Entry} instances.
+     *
+     * <p>
```
Comment (mtopolnik) [inline hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/connector/kafka/ReadKafkaP.java]: General note: we seem to have standardized on putting just one `<p>` on a line of its own to separate paragraphs. IDEA does it the way it's here now, but it's not a very convenient way to treat paragraph marks.

File: hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/connector/kafka/ReadKafkaP.java [hunk 2]
```diff
@@ -145,33 +155,62 @@ public void close() {
...
+     *     You can specify a partition to offset mapper to mark the start offset of each partition.
+     *     Any negative value as an offset will throw {@code IllegalArgumentException}
+     * </p>
```
Comment (mtopolnik) [inline hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/connector/kafka/ReadKafkaP.java]: I'm missing documentation that explains the meaning of the left and right operand of `BiFunction`.

File: hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/connector/kafka/ReadKafkaP.java [hunk 3]
```diff
@@ -189,10 +228,11 @@ public void init(@Nonnull Context context) {
             consumer.close();
         }
 
-        @Override @Nonnull
+        @Override
+        @Nonnull
```
Comment (mtopolnik) [inline hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/connector/kafka/ReadKafkaP.java]: In all other places we put both of these annotations on the same line.

File: hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/connector/kafka/ReadKafkaP.java [hunk 4]
```diff
@@ -202,35 +242,39 @@ public void init(@Nonnull Context context) {
...
         public List<Processor> get(int count) {
             Map<Integer, List<Integer>> processorToPartitions =
                     IntStream.range(0, ownedPartitions.size()).boxed()
-                             .map(i -> new SimpleImmutableEntry<>(i, ownedPartitions.get(i)))
-                             .collect(groupingBy(e -> e.getKey() % count,
-                                     mapping(Entry::getValue, toList())));
+                            .map(i -> new SimpleImmutableEntry<>(i, ownedPartitions.get(i)))
```
Comment (mtopolnik) [inline hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/connector/kafka/ReadKafkaP.java]: We align at the dot in these expressions. This is configurable in IDEA: **Preferences > Editor > Code Style > Java > Wrapping and Braces > Chained method calls > Align when multiline.**

File: hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/connector/kafka/ReadKafkaP.java (hunk 2 above)
Comment (gurbuzali) [inline hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/connector/kafka/ReadKafkaP.java]: `partitionId, currentOffset -> desiredOffset` I will add it to the docs.

File: hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/connector/kafka/ReadKafkaP.java (hunk 4 above)
Comment (gurbuzali) [inline hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/connector/kafka/ReadKafkaP.java]: 👍

File: hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/connector/kafka/ReadKafkaP.java (hunk 3 above)
Comment (gurbuzali) [inline hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/connector/kafka/ReadKafkaP.java]: Idea does it, do you know any configuration for that too ?

File: hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/connector/kafka/ReadKafkaP.java [hunk 5]
```diff
@@ -145,33 +155,62 @@ public void close() {
...
+     * @param brokerConnectionString kafka broker address
+     * @param deserializeKey         function for deserializing keys
+     * @param deserializeValue       function for deserializing values
+     * @param partitionOffsetMapper  function for mapping partitionOffsets
+     */
+    public static <K, V> ProcessorMetaSupplier readKafka(String zkAddress, String groupId, String topicId,
+                                                         String brokerConnectionString,
+                                                         Function<byte[], K> deserializeKey,
+                                                         Function<byte[], V> deserializeValue,
+                                                         BiFunction<Integer, Long, Long> partitionOffsetMapper) {
```
Comment (cangencer) [inline hazelcast-jet-kafka/src/main/java/com/hazelcast/jet/connector/kafka/ReadKafkaP.java]: the list of parameters is getting really long now. Maybe we should just accept the Properties object?

This seems to be the typical case when creating KafkaConsumer: https://www.confluent.io/blog/tutorial-getting-started-with-the-new-apache-kafka-0-9-consumer-client/

Also it has support for some custom deserializer, from the `Deserializer` interface.

Could you do some more research around in this area?

**Pull request closed without merging**